deploying to kubernetes guide (#23978)
dagster-io/dagster: docs/docs-beta/docs/guides/deployment/kubernetes.md

- **Familiarity with [Docker](https://docs.docker.com/)**, and:
  - **To have Docker installed**. [Docker installation guide](https://docs.docker.com/engine/install/)
  - **Access to a Docker image registry**, such as Amazon Web Services ECR or DockerHub. If you're following along on your local machine, this isn't required.
- **Familiarity with [Kubernetes](https://kubernetes.io/docs/home/)**, and:
  - **To have `kubectl` installed**. [Kubernetes installation guide](https://kubernetes.io/docs/tasks/tools/)
  - **An existing Kubernetes cluster**. To follow along on your local machine, [install Docker Desktop](https://docs.docker.com/desktop/kubernetes/) and turn on the included Kubernetes server.
- **Familiarity with [Helm](https://helm.sh/docs/)**, and:
  - **To have Helm 3 installed**. [Helm installation guide](https://helm.sh/docs/intro/install/)
- A Dagster project to deploy. You can also use the [example project](/todo):
  ```bash
  dagster project from-example --example deploy_k8s_beta --name deploy_k8s_beta
  ```

</details>

## Step 1: Write and build a Docker image containing your Dagster project
### Step 1.1: Write a Dockerfile

> Next, you'll build a Docker image that contains your Dagster project and all of its dependencies. The Dockerfile should:

Reviewer: I think we should mention where exactly to put this Dockerfile relative to this project.

Author: what is the recommendation? I put mine at the top level, but i dont know if that's best practices

Reviewer: Yes, also must be top level for the COPY command to work.

```bash
docker build . -t iris_analysis:1
```
This builds the Docker image from Step 1.1 and gives it the name `iris_analysis` and tag `1`. You can set custom values for both the name and the tag. We recommend that each time you rebuild your Docker image, you assign a new value for the tag to ensure that the correct image is used when running your code.

If you are using a Docker image registry, push the image to your registry. If you are following along on your local machine, you can skip this command.

```bash
docker push iris_analysis:1
```

If you are pushing your image to an image registry, you can find more information about this process in your registry's documentation:
- [Amazon ECR](https://docs.aws.amazon.com/AmazonECR/latest/userguide/docker-push-ecr-image.html)
- [DockerHub](https://docs.docker.com/docker-hub/quickstart/#step-5-build-and-push-a-container-image-to-docker-hub-from-your-computer)

## Step 2: Configure `kubectl` to point at a Kubernetes cluster
Before you can deploy Dagster, you need to configure `kubectl` to develop against the Kubernetes cluster where you want Dagster to be deployed.

> If you are using Docker Desktop and the included Kubernetes server, you will need to create a context first. If you already have a Kubernetes cluster and context created for your Dagster deployment you can skip running this command.

Reviewer: I feel we can assume the user knows the basics of kubernetes here and dont need special instructions for docker desktop. We should just have them run `use-context` below.

To install the Dagster Helm charts, run the following command:

```bash
helm repo add dagster https://dagster-io.github.io/helm
```

If you have previously added the Dagster Helm charts, run the following command to update the repository:

```bash
helm repo update
```

## Step 4: Configure the Helm chart for your deployment

You will need to modify some values in Dagster's Helm chart to deploy your Dagster project.

### Step 4.1: Copy default Helm chart values into values.yaml

Run the following command to copy the values installed from the published Helm charts:

```bash
helm show values dagster/dagster > values.yaml
```

### Step 4.2: Modify the `values.yaml` file for your deployment
The `values.yaml` file contains configuration options you can set for your deployment. Different configuration options are explained in [inline comments in `values.yaml`](https://artifacthub.io/packages/helm/dagster/dagster?modal=values).

To deploy your project, you'll need to set the following options:
- `dagster-user-deployments.deployments.name`, which should be a unique name for your deployment

> - `dagster-user-deployments.deployments.image.name` and `dagster-user-deployments.deployments.image.tag`, which should be set to match the Docker image from Step 1

Reviewer: Maybe include the literal `iris_analysis:1`?

If you are following this guide on your local machine, you will also need to set `pullPolicy: IfNotPresent`. This will use the local version of the image built in Step 1. However, in production use cases when your Docker images are pushed to image registries, this value should remain `pullPolicy: Always`.

<CodeExample filePath="guides/deployment/kubernetes/minimal_values.yaml" language="yaml" title="Minimal changes to make to values.yaml" />

In this example, the image `name` and `tag` are set to `iris_analysis` and `1` to match the image that was pushed in Step 1. To run the gPRC server, the path to the Dagster project needs to be specified, so `--python-file` and `/iris_analysis/definitions.py` are set for `dagsterApiGrpcArgs`.

## Step 5: Install the Helm chart
Now that you have modified the Helm `values.yaml` file, you can install the changes in your Kubernetes cluster.

Run the following command to install the Helm chart and create a [release](https://helm.sh/docs/intro/using_helm/#three-big-concepts).

```bash
helm upgrade --install dagster dagster/dagster -f /path/to/values.yaml
```

> :::note

Reviewer: Older versions are an advanced topic, and should be left out in the interest of brevity.

- Familiarity with [Docker](https://docs.docker.com/)
- Familiarity with [Kubernetes](https://kubernetes.io/docs/home/)
- Familiarity with [Helm](https://helm.sh/docs/)
- A Dagster project to deploy. You can also use the [example project](/todo). You can download the example project using the command:
```bash
dagster project from-example --example deploy_k8s_beta --name assets_dbt_python
```
- To have Docker installed. [Docker installation guide](https://docs.docker.com/engine/install/)
- To have `kubectl` installed. [Kubernetes installation guide](https://kubernetes.io/docs/tasks/tools/)
- To have a Kubernetes cluster created. If you would like to follow along with this guide on your local machine, you can install Docker Desktop and turn on the included Kubernetes server [Docker Desktop and Kubernetes guide](https://docs.docker.com/desktop/kubernetes/)
- Access to a Docker image registry, such as Amazon Web Services ECR or DockerHub. If you are following along to the guide on your local machine, you do not need access to a Docker image registry.
- To have Helm 3 installed. [Helm installation guide](https://helm.sh/docs/intro/install/)

</details>

> ## Step 1: Get your Dagster project ready to deploy

Reviewer 2: Thoughts on shortening this? Ex: `Step 1: Prep your Dagster project` or even `Prepare`

You will need to add a `workspace.yaml` file to your Dagster project to be ready to deploy with Kubernetes. If you are using the [example project](/todo), you will already have this file.

The `workspace.yaml` file tells Dagster where to find the `Definitions` object in your project. This file should be at the LOCATION of your project.

<CodeExample filePath="guides/deployment/kubernetes/workspace.yaml" language="yaml" title="Example workspace.yaml" />

> For example, in the sample `iris_analysis` project, the `Definitions` object is found in the `iris_analysis/__init__.py` file. The `workspace.yaml` tells Dagster to load this project and give it the name `iris_analysis`.

Reviewer 2: > is found in the `iris_analysis/__init__.py`

DevSuccess has start updating everything to be in `definitions.py` instead of `__init__.py`, as part of a new project structuring best practice. We should probably follow that guidance here to be consistent.

Author: good to know! will do

You can also point to a Python module in the `workspace.yaml` file. To learn more about writing a `workspace.yaml` see the [Workspace](/todo) guide.

## Step 2: Write and build a Docker image containing your Dagster project
### Step 2.1: Write a Dockerfile

> You will need to build a Docker image that contains your Dagster project and all of its dependencies. The Dockerfile should copy your Dagster project and the `workspace.yaml` file created in Step 1 into the image and install `dagster`, `dagster-postgres`, and `dagster-k8s`, along with any other libraries your project depends on. Finally, ensure that port 80 is exposed. This will be used to set up port-forwarding later.

Reviewer 2: I think this is a lot of dense information - thoughts on breaking it up a little? Ex:

```suggestion
Next, you'll build a Docker image that contains your Dagster project and all of its dependencies. The Dockerfile should:

1. Copy your Dagster project and the `workspace.yaml` file into the image
2. Install `dagster`, `dagster-postgres`, and `dagster-k8s`, along with any other libraries your project depends on. The example project has a `pandas` dependency, so it's included in `pip install` in the following example `Dockerfile`.
3. Expose port `80`, which we'll use to set up port-forwarding later in the guide.
```

Reviewer 2: Also:

> The Dockerfile should copy your Dagster project and the `workspace.yaml` file

Do we need to specifically include `workspace.yaml` here? Isn't it part of the Dagster project?

Author: What about 
> Copy your Dagster project, including the `workspace.yaml` file created in Step 1, into the image.

Reviewer 2: Works for me!

If you are using a Docker image registry, push the image to your registry. If you are following along on your local machine, you can skip this command.

```bash
docker push iris_analysis:1
```

TODO - specific instructions if pushing to an image registry?

## Step 3: Configure `kubectl` to point at a Kubernetes cluster
Before you can deploy Dagster, you need to configure `kubectl` to develop against the Kubernetes cluster where you want Dagster to be deployed.

If you are using Docker Desktop and the included Kubernetes server, you will need to create a context first. If you already have a Kubernetes cluster and context created for your Dagster deployment you can skip running this command.
```bash
kubectl config set-context dagster --namespace default --cluster docker-desktop --user=docker-desktop
```

Ensure that `kubectl` is using the correct context by running:
```bash
kubectl config use-context <context-name>
```
Where `<context-name>` is the name of the context you want to use. For example, if you ran the preceding `kubectl config set-context` command, you will run
```bash
kubectl config use-context dagster
```

## Step 4: Add the Dagster Helm chart repository

> Dagster publishes [Helm charts](https://artifacthub.io/packages/helm/dagster/dagster) for deploying Dagster. New Helm charts are published for each version of Dagster. You should use the Helm chart version that matches the version of Dagster you have installed.

Reviewer 2: ```suggestion
Dagster publishes [Helm charts](https://artifacthub.io/packages/helm/dagster/dagster) for deploying Dagster, with a new chart for each Dagster version. Use the Helm chart version that matches the version of Dagster you have installed.
```

Trying to make this read a little better, but I'm not sure I accomplished it with how many times I said `Dagster` 😅

Author: i think this is better

Author: however i do think i'm missing something with the version thing and that it needs to be used in an earlier command. I'll wait to get the tech review to figure out the right spot for `--version` and then add the note about matching your dagster version there

To install the Dagster Helm charts, run the following command:

```bash
helm repo add dagster https://dagster-io.github.io/helm
```

If you have previously added the Dagster Helm charts, run the following command to update the repository:

```bash
helm repo update
```

## Step 5: Configure the Helm chart for your deployment

You will need to modify some values in Dagster's Helm chart to deploy your Dagster project.

> ### Step 5.1: Copy the default Helm chart values into a `values.yaml`

Reviewer 2: ```suggestion
### Step 5.1: Copy default Helm chart values into a values.yaml
```

No formatting in headings pls

> Run the following command to copy the values installed from the published Helm charts so that you can modify them:

Reviewer 2: ```suggestion
Run the following command to copy the values installed from the published Helm charts:
```

Suggesting we drop the last few words, since this section intro already described what the user will be doing

```bash
helm show values dagster/dagster > values.yaml
```
TODO - where to copy this file to?

### Step 5.2: Modify the `values.yaml` file for your deployment
The `values.yaml` file contains configuration options you can set for your deployment. Different configuration options are explained in inline comments in `values.yaml`, and you can learn more about the different configuration options in [The title of a guide explaining it](/todo).

> The minimal configuration options you need to set to deploy your project are the `deployments.name`, `deployments.image`, and `deployments.dagsterApiGrpcArgs` values. `deployments.name` should be a unique name for your deployment, and `deployments.image` should be set to match the Docker image you built and pushed in Step 2. `dagsterApiGrpcArgs` should be set to NEED HELP WITH HOW TO EXPLAIN THIS.

Reviewer 2: Thoughts on putting this into a list?

```suggestion
To deploy your project, you'll need to set the following options:

- `deployments.name`, which should be a unique name for your deployment
- `deployments.image`, which should be set to match the Docker image from Step 2
- `deployments.dagsterApiGrpcArgs`, which should be set to NEED HELP WITH HOW TO EXPLAIN THIS.
```

If you are following this guide on your local machine, you will also need to set `pullPolicy: IfNotPresent`. This will use the local version of the image built in Step 2. However, in production use cases when your Docker images are pushed to image registries, this value should remain `pullPolicy: Always`.

<CodeExample filePath="guides/deployment/kubernetes/minimal_values.yaml" language="yaml" title="Minimal changes to make to values.yaml" />

## Step 6: Install the Helm chart
Now that you have modified the Helm `values.yaml` file, you can install the changes in your Kubernetes cluster.

Run the following command to install the Helm chart and create a release (TODO i dont really know what create a release means in this context, just took it from the existing guide)

```bash
helm upgrade --install dagster dagster/dagster -f /path/to/values.yaml
```
I think it would be nice to explain the various components of this command and what to set them to, but i dont know what the command is really doing

:::note

> If you are running an older version of Dagster, pass the --version flag to `helm upgrade` with the version of Dagster you are running. For example, if you are running `dagster==1.7.4` you'll run the command `helm upgrade --install dagster dagster/dagster -f /path/to/values.yaml --version 1.7.4`

Reviewer 2: ```suggestion
If you are running an older version of Dagster, pass the `--version` flag to `helm upgrade` with the version of Dagster you are running. For example, if you are running `dagster==1.7.4` you'll run the command `helm upgrade --install dagster dagster/dagster -f /path/to/values.yaml --version 1.7.4`
```

:::

The `helm upgrade` command will launch several pods in your Kubernetes cluster. You can check the status of the pod with the command:

```bash
kubectl get pods
```

It may take a few minutes before all pods are in a `RUNNING` state. If the `helm upgrade` was successful, you should see a `kubectl get pods` output similar to this:

```bash
$ kubectl get pods
NAME                                                              READY   STATUS      AGE
dagster-daemon-5787ccc868-nsvsg                                   1/1     Running     3m41s
dagster-dagit-7c5b5c7f5c-rqrf8                                    1/1     Running     3m41s
dagster-dagster-user-deployments-iris-analysis-564cbcf9f-fbqlw    1/1     Running     3m41s
dagster-postgresql-0                                              1/1     Running     3m41s
```

<details>
  <summary>Debugging failed pods</summary>

If one of the pods is in an error state, you can view the logs using the command

```bash
kubectl logs <pod-name>
```

> For example, if the pod `dagster-dagster-user-deployments-iris-analysis-564cbcf9f-fbqlw` is in a `CrashLoopBackOff` state, the logs can be viewed with the command

Reviewer 2: I like this bit, but what do you think about choosing a different pod for this example? The name is so long and I think a short one would still get the point across.

```
kubectl logs dagster-dagster-user-deployments-iris-analysis-564cbcf9f-fbqlw
```

</details>

TODO - maybe explain what each of these pods are?

## Step 7: Connect to your Dagster deployment and materialize your assets

### Step 7.1: Start port-forwarding to the webserver pod

> Run the following command to set up port forwarding to the webserver pod"

Reviewer 2: ```suggestion
Run the following command to set up port forwarding to the webserver pod:
```

```bash
DAGSTER_WEBSERVER_POD_NAME=$(kubectl get pods --namespace default \
  -l "app.kubernetes.io/name=dagster,app.kubernetes.io/instance=dagster,component=dagster-webserver" \
  -o jsonpath="{.items[0].metadata.name}")
kubectl --namespace default port-forward $DAGSTER_WEBSERVER_POD_NAME 8080:80
```

This command gets the full name of the `webserver` pod from the output of `kubectl get pods`, and then sets up port forwarding with the `kubectl port-forward` command.

### Step 7.2: Visit your Dagster deployment
The webserver has been port-forwarded to `8080`, so you can visit the Dagster deployment by going to [http://127.0.0.1:8080](http://127.0.0.1:8080). You should see the Dagster landing page

TODO SCREENSHOT

### Step 7.3: Materialize an asset

> From the Dagster UI you can materialize an asset by clicking the "Materialize" button. Dagster will start a Kubernetes job to materialize the asset. You can introspect on the Kubernetes cluster to see this job:

Reviewer 2: ```suggestion
In the Dagster UI, click the **Materialize** button to materialize an asset. Dagster will start a Kubernetes job to materialize the asset. You can introspect on the Kubernetes cluster to see this job:
```

Some light suggestions here, but where in the UI should the user find & click **Materialize**? I don't think it needs to be overly descriptive, but just pointing folks to a page (ex: "The Asset catalog") would be enough

---
title: "Deploy to Kubernetes"
sidebar_position: 21
---

> Deploying Dagster to a Kubernetes cluster is a popular option for maintaining a production Dagster deployment. Dagster uses Helm, a package manager for Kubernetes applications, to help manage deploying Dagster to a Kubernetes cluster.

Reviewer 3: > a production Dagster deployment

I think we should position this (and other OSS deployment) docs as reference/examples for how to run the Dagster-specific parts of a productionized Dagster deployment.

Productionizing it for real will probably require additional decisions around networking, access control, running persistent data stores, monitoring, etc. that are all out of scope for the Dagster project and have too large of a decision matrix for us to fully capture.

Reviewer 4: I think I agree.  The goal of this guide shouldnt be to teach you how to build production systems, but rather, what the components of a Dagster production deployment are.

Author: ok i can reframe this intro. Do you think there are other systemic changes that need to be made to the doc to position it this way?

Author: is a doc that discusses thing like networking, access control data stores, etc something that we plan to do?

<details>
  <summary>Prerequisites</summary>

To follow the steps in this guide, you'll need:

- **Familiarity with [Docker](https://docs.docker.com/)**, and:
  - **To have Docker installed**. [Docker installation guide](https://docs.docker.com/engine/install/)
  - **Access to a Docker image registry**, such as Amazon Web Services ECR or DockerHub. If you're following along on your local machine, this isn't required.
- **Familiarity with [Kubernetes](https://kubernetes.io/docs/home/)**, and:
  - **To have `kubectl` installed**. [Kubernetes installation guide](https://kubernetes.io/docs/tasks/tools/)
  - **An existing Kubernetes cluster**. To follow along on your local machine, [install Docker Desktop](https://docs.docker.com/desktop/kubernetes/) and turn on the included Kubernetes server.
- **Familiarity with [Helm](https://helm.sh/docs/)**, and:
  - **To have Helm 3 installed**. [Helm installation guide](https://helm.sh/docs/intro/install/)
- A Dagster project to deploy. You can also use the [example project](/todo):
  ```bash
  dagster project from-example --example deploy_k8s_beta --name deploy_k8s_beta
  ```

</details>

## Step 1: Prepare your Dagster project

> You will need to add a `workspace.yaml` file to your Dagster project to be ready to deploy with Kubernetes. If you are using the [example project](/todo), you will already have this file.

Reviewer 5: you might need a workspace.yaml to run locally, but you don't need it to deploy to kubernetes. the helm chart creates a workspace.yaml that the k8s cluster uses based on the information under dagster-user-deployments

Author: oh sick, i didn't know that!

To install the Dagster Helm charts, run the following command:

```bash
helm repo add dagster https://dagster-io.github.io/helm
```

If you have previously added the Dagster Helm charts, run the following command to update the repository:

```bash
helm repo update
```

## Step 5: Configure the Helm chart for your deployment

You will need to modify some values in Dagster's Helm chart to deploy your Dagster project.

### Step 5.1: Copy default Helm chart values into values.yaml

Run the following command to copy the values installed from the published Helm charts:

```bash
helm show values dagster/dagster > values.yaml
```
TODO - where to copy this file to?

### Step 5.2: Modify the `values.yaml` file for your deployment
The `values.yaml` file contains configuration options you can set for your deployment. Different configuration options are explained in inline comments in `values.yaml`, and you can learn more about the different configuration options in [The title of a guide explaining it](/todo).

To deploy your project, you'll need to set the following options:

> - `deployments.name`, which should be a unique name for your deployment

Reviewer 5: believe this is `dagster-user-deployments.deployments` throughout, not just `deployments`

- `deployments.image.name` and `deployments.image.tag`, which should be set to match the Docker image from Step 2

> - `deployments.dagsterApiGrpcArgs`, which should be set to NEED HELP WITH HOW TO EXPLAIN THIS.

Reviewer 5: https://docs.dagster.io/concepts/code-locations/workspace-files#running-your-own-grpc-server shows the arguments here - it's the arguments you would pass into the `dagster api grpc` command to spin up a code server. The helm chart does this for you. It's unfortunate that a user has to think about the word "grpc" here.

Author: @gibsondan what do you think of the description i ended up with?

If you are following this guide on your local machine, you will also need to set `pullPolicy: IfNotPresent`. This will use the local version of the image built in Step 2. However, in production use cases when your Docker images are pushed to image registries, this value should remain `pullPolicy: Always`.

<CodeExample filePath="guides/deployment/kubernetes/minimal_values.yaml" language="yaml" title="Minimal changes to make to values.yaml" />

## Step 6: Install the Helm chart
Now that you have modified the Helm `values.yaml` file, you can install the changes in your Kubernetes cluster.

> Run the following command to install the Helm chart and create a release (TODO i dont really know what create a release means in this context, just took it from the existing guide)

Reviewer 5: release is a Helm concept: https://helm.sh/docs/intro/using_helm/#three-big-concepts

```bash
helm upgrade --install dagster dagster/dagster -f /path/to/values.yaml
```
I think it would be nice to explain the various components of this command and what to set them to, but i dont know what the command is really doing

:::note

> If you are running an older version of Dagster, pass the `--version` flag to `helm upgrade` with the version of Dagster you are running. For example, if you are running `dagster==1.7.4` you'll run the command `helm upgrade --install dagster dagster/dagster -f /path/to/values.yaml --version 1.7.4`

Reviewer 5: this is specifically the version of dagster that the system components will use like the webserver and the daemon (you can have your code running on earlier versions)

Author: do they need to pass a version flag when copying the `values.yaml` too then?

---
title: "Deploy to Kubernetes"
sidebar_position: 21
---

> TODO - INTRODUCTION

Author: Still need to write the intro

## What you'll learn

- How to use Kubernetes and Helm to deploy Dagster

<details>
  <summary>Prerequisites</summary>

To follow the steps in this guide, you'll need:

- Familiarity with [Docker](https://docs.docker.com/)
- Familiarity with [Kubernetes](https://kubernetes.io/docs/home/)
- Familiarity with [Helm](https://helm.sh/docs/)

> - A Dagster project to deploy. You can also use the example project [here](/todo). Would be great to have a command they can use to clone the project

Author: the existing guide has a corresponding example project that you can download. I think it would be nice to do the same thing with the new project, but that could require some additional overhead for keeping the code snippets in this guide up to date with the code in the example project directory. thoughts?

Author: it ends up just being the `workspace.yaml` that needs to be maintained in two locations so i think it's worth having a project they can download. I'll make that

- To have Docker installed. [Docker installation guide](https://docs.docker.com/engine/install/)
- To have `kubectl` installed. [Kubernetes installation guide](https://kubernetes.io/docs/tasks/tools/)
- To have a Kubernetes cluster created. If you would like to follow along with this guide on your local machine, you can install Docker Desktop and turn on the included Kubernetes server [Docker Desktop and Kubernetes guide](https://docs.docker.com/desktop/kubernetes/)
- Access to a Docker image registery, such as Amazon Web Services ECR or DockerHub
- To have Helm 3 installed. [Helm installation guide](https://helm.sh/docs/intro/install/)

</details>

> ## Step 0a: Understand the Dagster deployment architecture

Author: The existing guide has a deployment architecture section https://docs.dagster.io/deployment/guides/kubernetes/deploying-with-helm#deployment-architecture

should i do something similar here? If the architecture is more general purpose than just kubernetes, i'd be pro having it as a separate guide/reference page and linking to it as a pre-requisite

Reviewer 2: I'd leave it out of this guide, since we want to focus on the how and less on the what. We can revisit it as another page at some point

Author: :relief

Do we need this part? Should it be a pre-req that we link to?

## Step 0b: Example project tour

> If you are deploying your own Dagster project, skip ahead to Step 1. However, if you are using the example project for this guide, this step will walk you through the contents of the project.

Author: debating whether it makes sense to go over the example project structure. It's a super simple project and i'm not sure going through the one file really adds anything

Author: now i'm thinking i'll put a readme in the example project that briefly goes over the project layout. then i can remove this section

The example project should contain the following files
```bash
TODO FILE TREE
```

The Dagster project can be found in `iris_analysis`. The project itself is found in the `iris_analysis/__init__.py` file. This file contains a single asset that downloads a dataset about iris flowers and logs the number of rows.

The example project also contains a `workspace.yaml` file and a `Dockerfile`. These files will be covered in the following steps.

## Step 1: Get your Dagster project ready to deploy
You will need to add a `workspace.yaml` file to your Dagster project to be ready to deploy with Kubernetes. If you are using the [example project](/todo), you will already have this file.

> The `workspace.yaml` file tells Dagster where to find the `Definitions` object in your project. This file should be at the LOCATION of your project.

Author: what's our current advice on where `workspace.yaml` files should live?

Reviewer 2: I thought they went at the same level as `pyproject.toml`?

Author: i biffed writing out my real question - what language are we using to tell people where `workspace.yaml` goes? like are we saying "in the top directory of your project"? "in the same directory as pyproject.toml"?

Reviewer 4: i would say same directory as pyproject.toml. since they could have a monorepo with many top directories

Reviewer 5: as per above this isn't needed on k8s - it will look in dagsterApiGrpcArgs (gross) in the helm chart for this

Author: i removed workspace!

<CodeExample filePath="guides/deployment/kubernetes/workspace.yaml" language="yaml" title="Example workspace.yaml" />
For example, in the sample `iris_analysis` project, the `Definitions` object is found in the `iris_analysis/__init__.py` file. The `workspace.yaml` tells Dagster to load this project and give it the name `iris_analysis`.

You can also point to a python module in the `workspace.yaml` file. To learn more about writing a `workspace.yaml` see the [Workspace](/todo) guide.

## Step 2: Write and build a Docker image containing your Dagster project

> ### Step 2.1: Write a Dockerfile

Author: Dockerfile is generally capitalized, so i'm going to add this to vale's ignore list unless told otherwise

Reviewer 6: That's fine, I believe Sandy's PR also adds it.

You will need to build a Docker image that contains your Dagster project and all of its dependencies. The Dockerfile should copy your Dagster project and the `workspace.yaml` file created in Step 1 into the image and install `dagster`, `dagster-postgres`, and `dagster-k8s`, along with any other libraries your project depends on. Finally, ensure that port 80 is exposed. This will be used to set up port-forwarding later.

<CodeExample filePath="guides/deployment/kubernetes/Dockerfile" language="docker" title="Example Dockerfile" />
The example project has a dependency on `pandas` so it is included in the `pip install` command.

### Step 2.2: Build and push a Docker image

To build your Docker image, run the following command from the directory where your Dockerfile is located:

```bash
docker build . -t iris_analysis:1
```

> This builds the Docker image from Step 2.1 and gives it the name `iris_analysis` and tag `1`. You can set custom values for both the name and the tag. We recommend that each time you rebuild your Docker image, you assign a new value for the tag to ensure that the correct image is used when running your code.

Author: for the tech reviewer: i'd like to spot check that this statement is correct

Reviewer 4: paging dr @jmsanders

Reviewer 3: That is correct. I also wonder how much prior knowledge we want to assume here? Or how much we want to offload some of the pre-req stuff to third party docs (specifically the Docker and/or K8s docs)?

I think it's fair to assume if you're going to manage your own open source deployment of Dagster on K8s, you:
- have basic familiar with containers
- have basic familiar with k8s
- or are comfortable reading up on those two things first

If you are using a Docker image registry, push the image to your registery. If you are following along on your local machine, you can skip this command.

```bash
docker push iris_analysis:1
```

> TODO - specific instructions if pushing to an image registry?

Author: Is there anything specific we should mention for users who will push to a registry?

Reviewer 5: It might make sense to just link to other docs here? Here's the one for ECR https://docs.aws.amazon.com/AmazonECR/latest/userguide/docker-push-ecr-image.html

```
Where `<context-name>` is the name of the context you want to use. For example, if you ran the `kubectl config set-context` command above, you will run
```bash
kubectl config use-context dagster
```

## Step 4: Add the Dagster Helm chart repository

Dagster publishes [Helm charts](https://artifacthub.io/packages/helm/dagster/dagster) for deploying Dagster. New Helm charts are published for each version of Dagster. You should use the Helm chart version that matches the version of Dagster you have installed.

To install the Dagster Helm charts, run the following command:

```bash
helm repo add dagster https://dagster-io.github.io/helm
```

If you have previously added the Dagster Helm charts, run the following command to update the repository:

```bash
helm repo update
```

## Step 5: Configure the Helm chart for your deployment

You will need to modify some values in Dagster's Helm chart to deploy your Dagster project.

### Step 5.1: Copy the default Helm chart values into a `values.yaml`

Run the following command to copy the values installed from the published Helm charts so that you can modify them:

```bash
helm show values dagster/dagster > values.yaml
```

> TODO - where to copy this file to?

Author: do we have a preference or advice on where to make the `values.yaml` file?

Reviewer 5: I think a lot of people would check this into source control

### Step 5.2: Modify the `values.yaml` file for your deployment

> The `values.yaml` file contains configuration options you can set for your deployment. There are comments in the `values.yaml` file explaining these options, and you can learn more about them [here](/todo).

Author: Not sure if a doc exists that documents the various options you can set in the `values.yaml`

Reviewer 4: I think I saw a post where Jordan found a way to generate them. For now a /todo is fine

Reviewer 3: https://artifacthub.io/packages/helm/dagster/dagster?modal=values has all the possible values.

The thing I was generating was documentation around dagster.yaml config (and in my case, only for Dagster+ - although we could probably extend it to use elsewhere in the docs site). Similar, but not quite the same:

- Helm values customize our Helm templates. Our Helm templates describe various Kubernetes resources to be created with the Kubernetes API (deployments, services, configmaps, etc.)
- dagster.yaml configures the `DagsterInstance` class.
- This Helm template creates a [configmap](https://kubernetes.io/docs/concepts/configuration/configmap/) that [contains a dagster.yaml](https://github.com/dagster-io/dagster/blob/1c47efad19fd3027704fc4f5d8a7c9da18bdb699/helm/dagster/templates/configmap-instance.yaml#L11). So some of the options in this particular values.yaml map onto options in a dagster.yaml that it creates for us.

But yeah, I'd just link directly out to the Helm template docs in artifacthub here.

> The minimal configuration options you need to set in order to deploy your project are the `deployments.name`, `deployments.image`, and `deployments.dagsterApiGrpcArgs` values. `deployments.name` should be a unique name for your deployment, and `deployments.image` should be set to match the Docker image you built and pushed in Step 2. `dagsterApiGrpcArgs` should be set to NEED HELP WITH HOW TO EXPLAIN THIS.

Author: I believe this is the minimal set of options you need to set (at least it worked for me locally)

Also could use some help with explaining what to set the `dagsterApiGrpcArgs` values to. Like how should a user decide what to set these as? it looks to me like it should follow the `workspace.yaml` file but i dont know if that's true in all cases

If you are following this guide on your local machine, you will also need to set `pullPolicy: IfNotPresent`. This will use the local version of the image built in Step 2. However, in production use cases when your Docker images are pushed to image registries, this value should remain `pullPolicy: Always`.

<CodeExample filePath="guides/deployment/kubernetes/minimal_values.yaml" language="yaml" title="Minimal changes to make to values.yaml" />

## Step 6: Install the Helm chart
Now that you have modified the Helm `values.yaml` file, you can install the changes in your Kubernetes cluster.

> Run the following command to install the Helm chart and create a release (TODO i dont really know what create a release means in this context, just took it from the existing guide)

Author: could use some help with vocabulary here. I took this line from the existing guide, but i don't really know what it means. Maybe to someone with helm familiarity this is a perfectly explanatory sentence

```bash
helm upgrade --install dagster dagster/dagster -f /path/to/values.yaml
```

> I think it would be nice to explain the various components of this command and what to set them to, but i dont know what the command is really doing

Author: Todo - potentially explain what the different components of this command mean. Again, if this is self explanatory to a helm user, I can skip it

:::note

> If you are running an older version of Dagster, pass the --version flag to `helm upgrade` with the version of Dagster you are running. For example, if you are running `dagster==1.7.4` you'll run the command `helm upgrade --install dagster dagster/dagster -f /path/to/values.yaml --version 1.7.4`

Author: for the tech reviewer: I want to confirm that this is the right command to pass `--version` to. The existing guide is not very specific

:::

The `helm upgrade` command will launch several pods in your Kubernetes cluster. You can check the status of the pod with the command:

```bash
kubectl get pods
```

It may take a few minutes before all pods are in a `RUNNING` state. If the `helm upgrade` was successful, you should see a `kubectl get pods` output similar to this:

```bash
$ kubectl get pods
NAME                                                              READY   STATUS      AGE
dagster-daemon-5787ccc868-nsvsg                                   1/1     Running     3m41s
dagster-dagit-7c5b5c7f5c-rqrf8                                    1/1     Running     3m41s
dagster-dagster-user-deployments-iris-analysis-564cbcf9f-fbqlw    1/1     Running     3m41s
dagster-postgresql-0                                              1/1     Running     3m41s
```

<details>
  <summary>Debugging failed pods</summary>

> If one of the pods is in an error state, you can view the logs using the command

Author: Is this debugging section actually useful? for someone like me who hasn't used k8s in several year, the command to see logs is nice to have right there, but idk if that  would apply to the general reader of this doc

Reviewer 2: I actually like it since it's in a toggle that I can ignore if I don't need it.

```bash
kubectl logs <pod-name>
```

For example, if the pod `dagster-dagster-user-deployments-iris-analysis-564cbcf9f-fbqlw` is in a `CrashLoopBackOff` state, the logs can be viewed with the command

```
kubectl logs dagster-dagster-user-deployments-iris-analysis-564cbcf9f-fbqlw
```

</details>

> TODO - maybe explain what each of these pods are?

Author: worth explaining what each of the pods are?

```bash
DAGSTER_WEBSERVER_POD_NAME=$(kubectl get pods --namespace default \
  -l "app.kubernetes.io/name=dagster,app.kubernetes.io/instance=dagster,component=dagster-webserver" \
  -o jsonpath="{.items[0].metadata.name}")
kubectl --namespace default port-forward $DAGSTER_WEBSERVER_POD_NAME 8080:80
```

This command gets the full name of the `webserver` pod from the output of `kubectl get pods`, and then sets up port forwarding with the `kubectl port-forward` command.

### Step 7.2: Visit your Dagster deployment
The webserver has been port-forwarded to `8080`, so you can visit the Dagster deployment by going to [http://127.0.0.1:8080](http://127.0.0.1:8080). You should see the Dagster landing page

TODO SCREENSHOT

### Step 7.3: Materialize an asset
From the Dagster UI you can materialize an asset by clicking the "Materialize" button. Dagster will start a Kubernetes job to materialize the asset. You can introspect on the Kubernetes cluster to see this job:

```bash
$ kubectl get jobs
NAME                                               COMPLETIONS   DURATION   AGE
dagster-run-5ee8a0b3-7ca5-44e6-97a6-8f4bd86ee630   1/1           4s         11s
```

## Next steps

> - Forwarding Dagster logs from a Kubernetes deployment to AWS, Azure, GCP

Author: this is just me jotting down ideas of things we could link to. Other ideas
- guide on setting up your project to do k8s-job-per-step
- k8s config in tags

---
title: "Deploy to Kubernetes"
sidebar_position: 21

> ---

Author: todo - find a good place to put the links to our helm repo and docker image registry

- How to access and customize the Dagster-provided Helm charts
- How to apply the configuration in the Helm charts to your Kubernetes cluster to deploy your Dagster project
- How to connect to your deployed Dagster project

<details>
  <summary>Prerequisites</summary>

To follow the steps in this guide, you'll need:

- Familiarity with [Docker](https://docs.docker.com/)
- Familiarity with [Kubernetes](https://kubernetes.io/docs/home/)
- Familiarity with [Helm](https://helm.sh/docs/)
- A Dagster project to deploy. You can also use the [example project](/todo). You can download the example project using the command:
```bash
dagster project from-example --example deploy_k8s_beta --name assets_dbt_python
```
- To have Docker installed. [Docker installation guide](https://docs.docker.com/engine/install/)
- To have `kubectl` installed. [Kubernetes installation guide](https://kubernetes.io/docs/tasks/tools/)
- To have a Kubernetes cluster created. If you would like to follow along with this guide on your local machine, you can install Docker Desktop and turn on the included Kubernetes server [Docker Desktop and Kubernetes guide](https://docs.docker.com/desktop/kubernetes/)

> - Access to a Docker image registry, such as Amazon Web Services ECR or DockerHub. If you are following along to the guide on your local machine, you do not need access to a Docker image registry.

Reviewer 2: This list scared me when I first expanded the toggle 😅 I think we can condense things a bit - wdyt?

````suggestion
- **Familiarity with [Docker](https://docs.docker.com/)**, and:
  - **To have Docker installed**. Refer to the [Docker installation guide](https://docs.docker.com/engine/install/).
  - **Access to a Docker image registry**, such as Amazon Web Services ECR or DockerHub. If you're following along on your local machine, this isn't required.
- **Familiarity with [Kubernetes](https://kubernetes.io/docs/home/)**, and:
  - **To have `kubectl` installed**. Refer to the [Kubernetes installation guide](https://kubernetes.io/docs/tasks/tools/).
  - **An existing Kubernetes cluster**. To follow along on your local machine, [install Docker Desktop](https://docs.docker.com/desktop/kubernetes/) and turn on the included Kubernetes server
- **Familiarity with [Helm](https://helm.sh/docs/)**
- **A Dagster project to deploy**. You can also use the [example project](/todo):
   ```bash
   dagster project from-example --example deploy_k8s_beta --name assets_dbt_python
   ```
````

Author: so fair. it's a really long list! I like the sub-bullets approach

To install the Dagster Helm charts, run the following command:

```bash
helm repo add dagster https://dagster-io.github.io/helm
```

If you have previously added the Dagster Helm charts, run the following command to update the repository:

```bash
helm repo update
```

## Step 5: Configure the Helm chart for your deployment

You will need to modify some values in Dagster's Helm chart to deploy your Dagster project.

### Step 5.1: Copy the default Helm chart values into a `values.yaml`

Run the following command to copy the values installed from the published Helm charts so that you can modify them:

```bash
helm show values dagster/dagster > values.yaml
```
TODO - where to copy this file to?

### Step 5.2: Modify the `values.yaml` file for your deployment
The `values.yaml` file contains configuration options you can set for your deployment. There are comments in the `values.yaml` file explaining these options, and you can learn more about them [here](/todo).

> The minimal configuration options you need to set to deploy your project are the `deployments.name`, `deployments.image`, and `deployments.dagsterApiGrpcArgs` values. `deployments.name` should be a unique name for your deployment, and `deployments.image` should be set to match the Docker image you built and pushed in Step 2. `dagsterApiGrpcArgs` should be set to NEED HELP WITH HOW TO EXPLAIN THIS.

Author: I'm not sure how to explain where to get the values for `dagsterApiGrpcArgs`
